src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexer.Fixer.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    /// <summary>
+    /// CA1831, CA1832, CA1833: Use AsSpan or AsMemory instead of Range-based indexers when appropriate.
+    /// </summary>
+    public abstract class UseAsSpanInsteadOfRangeIndexerFixer : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(
+                UseAsSpanInsteadOfRangeIndexerAnalyzer.StringRuleId,
+                UseAsSpanInsteadOfRangeIndexerAnalyzer.ArrayReadOnlyRuleId,
+                UseAsSpanInsteadOfRangeIndexerAnalyzer.ArrayReadWriteRuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // The rules are mutually exclusive, so there can't be more than one for the same span:
+            var diagnostic = context.Diagnostics.Single();
```

Comment:
This is generally not recommended. You could theoretically have another third party analyzer generating the same diagnostic. We have had quite a few bugs in the past where we have customer reports of Single causing an exception.
```suggestion
            var diagnostic = context.Diagnostics.FirstOrDefault();
```